torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2491,6 +2491,42 @@ Tensor linalg_eig_backward(const std::vector<torch::autograd::Variable> &grads,
   }
 }
 
+// https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf, page 10
+// see also https://arxiv.org/pdf/1701.00392.pdf Eqs. (4.60) and (4.63)
+Tensor linalg_eig_jvp_eigenvalues(const Tensor& dA,
+                                  const Tensor& L,
+                                  const Tensor& V) {
+  const auto dAComplex = at::complex(dA, at::zeros(dA.size(-1), dA.options()));
+  const auto dVfactor = at::linalg_solve(V, at::matmul(dAComplex, V));
+  return at::diag(at::mul(at::eye(V.size(-1), V.options()), dVfactor));
+}
+
+Tensor linalg_eig_jvp_eigenvectors(const Tensor& dA,
+                                   const Tensor& L,
+                                   const Tensor& V) {
+  const auto dAComplex = at::complex(dA, at::zeros(dA.size(-1), dA.options()));
+  const auto dVfactor = at::linalg_solve(V, at::matmul(dAComplex, V));
+  const auto Lconj = L.conj();
+  const auto Econj = Lconj.unsqueeze(-2) - Lconj.unsqueeze(-1);
+  auto Fconj = at::pow(Econj, -1);
+  Fconj.diagonal(0, -2, -1).zero_();
+  return at::matmul(V, at::mul(Fconj, dVfactor));
+}
+
+std::tuple<Tensor, Tensor> linalg_eig_jvp(const Tensor& dA,
+                                          const Tensor& L,
+                                          const Tensor& V) {
+  const auto dAComplex = at::complex(dA, at::zeros(dA.size(-1), dA.options()));
+  const auto dVfactor = at::linalg_solve(V, at::matmul(dAComplex, V));
+  const auto Lconj = L.conj();
+  const auto Econj = Lconj.unsqueeze(-2) - Lconj.unsqueeze(-1);
+  const auto Fconj = at::pow(Econj, -1);
+  Fconj.diagonal(0, -2, -1).zero_();
+  return std::make_tuple(
+    at::diag(at::mul(at::eye(V.size(-1), V.options()), dVfactor)),
+    at::matmul(V, at::mul(Fconj, dVfactor)));
```

Comment:
Might be simpler to write `*` rather than `at::mul` and `/` rather than `at::div` for readability.

Reply:
A note on multiplying by `Fconj` / `eye`.

You might want to look how it's done in `linalg_eig_backward` and the tricks that are used there to avoid having to use mathematical tricks, but rather use advanced indexing. The problem with using masks, besides the efficiency, is that `NaN * 0 = NaN`, so you might be  propagating NaNs that you created in your intermediate computations, which is annoying. For example, isn't the first return equal to `dVfactor.diagonal(0, -2, -1)`?

https://github.com/pytorch/pytorch/blob/04fe4382ec8e6031d353530d6b8446ac9c9f507e/torch/csrc/autograd/FunctionsManual.cpp#L2433-L2492

Reply:
Yes indeed, it is equal to `dVfactor.diagonal(0, -2, -1)`. Fixed it :-)